### PR TITLE
[FLINK-17170][kinesis] Fix deadlock during stop-with-savepoint.

### PR DESCRIPTION
**Pull request by AHeise**
## What is the purpose of the change

During stop-with-savepoint cancel is called under lock in legacy sources. Thus, if the fetcher is trying to emit a record at the same time, it cannot obtain the checkpoint lock. This behavior leads to a deadlock while cancel awaits the termination of the fetcher.

## Brief change log

- The fix is to mostly rely on the termination inside the run method. As a safe-guard, close also awaits termination where close is always caused without lock.
- Adding ITCase that covers the behavior.

## Verifying this change

- Added FlinkKinesisITCase.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**yes** / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisITCase.java [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+import java.util.stream.IntStream;
+
+import static org.apache.flink.streaming.connectors.kinesis.config.ConsumerConfigConstants.STREAM_INITIAL_POSITION;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.lessThanOrEqualTo;
+import static org.junit.Assert.assertThat;
+
+/** IT cases for using Kinesis consumer/producer based on Kinesalite. */
+public class FlinkKinesisITCase {
```
Comment (dannycranmer) [inline flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisITCase.java:58]: Existing IT for Kinesis was added to the e2e module, inline with the Kafka variant:
- https://github.com/apache/flink/blob/master/flink-end-to-end-tests/flink-streaming-kinesis-test/src/test/java/org/apache/flink/streaming/kinesis/test/KinesisTableApiITCase.java

I agree they are best kept within this module, however we should make them consistent. We should at least raise a follow up Jira to tidy up

File: flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/KinesaliteContainer.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import static com.amazonaws.SDKGlobalConfiguration.ACCESS_KEY_ENV_VAR;
+import static com.amazonaws.SDKGlobalConfiguration.SECRET_KEY_ENV_VAR;
+
+/**
+ * A testcontainer based on Kinesalite.
+ *
+ * <p>Note that the more obvious localstack container with Kinesis took 1 minute to start vs 10
+ * seconds of Kinesalite.
+ */
+public class KinesaliteContainer extends GenericContainer<KinesaliteContainer> {
```
Comment (dannycranmer) [inline flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/KinesaliteContainer.java:41]: Duplicates https://github.com/apache/flink/blob/master/flink-end-to-end-tests/flink-streaming-kinesis-test/src/test/java/org/apache/flink/streaming/kinesis/test/containers/KinesaliteContainer.java#L24

**Review by dannycranmer — APPROVED** [review]
The fix LGTM. We now have 2x Kinesis IT that duplicate setup. I would suggest we either:
- Move the test from e2e test module to this module, or
- Create a follow up Jira to capture the tidy up

File: flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/KinesaliteContainer.java (hunk 2 above)
Comment (AHeise) [reply flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/KinesaliteContainer.java:41]: I moved that to the kinesis module.

**Review by dannycranmer — APPROVED** [review]
LGTM

**Comment by AHeise** [comment]
@flinkbot run azure

**Comment by AHeise** [comment]
@flinkbot run azure

**Comment by AHeise** [comment]
@flinkbot run azure

**Comment by AHeise** [comment]
Build was green when disabling FLINK-22566, so merging without disable commit.

**Pull request merged**